internal/postrs/bridge.go

```diff
@@ -0,0 +1,270 @@
+// #include <stdlib.h>
+// #include "prover.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"sync"
+	"unsafe"
+
+	"github.com/spacemeshos/post/config"
+	"github.com/spacemeshos/post/shared"
+)
+
+// gpuMtx is a mutual exclusion lock for calls to gpu functions. It is required
+// to prevent concurrent calls to the GPU library that are expected to cause a
+// crash.
+var gpuMtx sync.Mutex
+
+type DeviceClass int
+
+const (
+	ClassCPU     = DeviceClass((C.DeviceClass)(C.CPU))
+	ClassGPU     = DeviceClass((C.DeviceClass)(C.GPU))
+	ClassUnknown = DeviceClass((C.DeviceClass)(C.UNKNOWN))
+)
+
+type ComputeProvider struct {
+	ID         uint
+	Model      string
+	DeviceType DeviceClass
+}
+
+func (c DeviceClass) String() string {
+	switch c {
+	case ClassCPU:
+		return "CPU"
+	case ClassGPU:
+		return "GPU"
+	default:
+		return "Unknown"
+	}
+}
+
+var (
+	ErrDeviceBusy        = errors.New("device is busy")
+	ErrInvalidProviderID = errors.New("invalid provider ID")
+
+	ErrInvalidLabelsRange = errors.New("invalid labels range")
+	ErrOclError           = errors.New("OpenCL error")
+	ErrInvalidArgument    = errors.New("invalid argument")
+	ErrFetchProviders     = errors.New("failed to fetch providers")
+)
+
+const (
+	// LabelLength is the length of the label in bytes.
+	LabelLength = 16
+)
+
+func InitResultToError(retVal uint32) error {
+	switch retVal {
+	case C.InitializeOk:
+		return nil
+	case C.InitializeInvalidLabelsRange:
+		return ErrInvalidLabelsRange
+	case C.InitializeOclError:
+		return ErrOclError
+	case C.InitializeInvalidArgument:
+		return ErrInvalidArgument
+	case C.InitializeFailedToGetProviders:
+		return ErrFetchProviders
+	default:
+		return fmt.Errorf("unknown error")
+	}
+}
+
+func cScryptPositions(opt *option) ([]byte, *uint64, error) {
+	if !gpuMtx.TryLock() {
+		return nil, nil, ErrDeviceBusy
+	}
+	defer gpuMtx.Unlock()
+
+	cProviderId := C.uint32_t(*opt.providerID)
+	cN := C.uintptr_t(opt.n)
+	cCommitment := C.CBytes(opt.commitment)
+	defer C.free(cCommitment)
+	cDifficulty := C.CBytes(opt.vrfDifficulty)
+	defer C.free(cDifficulty)
+	init := C.new_initializer(cProviderId, cN, (*C.uchar)(cCommitment), (*C.uchar)(cDifficulty))
+	if init == nil {
+		return nil, nil, ErrInvalidProviderID
+	}
+
+	defer C.free_initializer(init)
+
+	outputSize := LabelLength * (opt.endPosition - opt.startPosition + 1)
+	cStartPosition := C.uint64_t(opt.startPosition)
+	cEndPosition := C.uint64_t(opt.endPosition)
+	cOutputSize := C.uint64_t(outputSize)
+	cOut := (C.calloc(cOutputSize, 1))
+	defer C.free(cOut)
+
+	var cIdxSolution C.uint64_t
+	retVal := C.initialize(init, cStartPosition, cEndPosition, (*C.uint8_t)(cOut), &cIdxSolution)
+	if err := InitResultToError(retVal); err != nil {
+		return nil, nil, err
+	}
+
+	var vrfNonce *uint64
+	if cIdxSolution != 0 { // TODO(mafa): since 0 could be a valid nonce, we should find a better way to indicate no solution (e.g. InitializeOk = no solution, InitializeOkPow = solution)
```

Comment:
It appears that `C.initialize` returns -1 for `cIdxSolution` if none was found, double check.